Design/nexusgeometrysaving.md

```diff
@@ -0,0 +1,74 @@
+
+It would be recommended to retain the switch that allowed us to write old-style processed nexus files but the default would be to have the new-style files written out. The switch would be exposed as a parameter on the algorithm.
+
+Reloading the files should work without any modifications, `LoadInstrument` already does the right thing.
+
+**Pros**
+
+1. Users do not need to be aware that this change is happening - they have enough to think about
+1. Provides a useful modernisation update to Mantid across the facilities
+
+**Cons**
+
+1. Potential for breaking changes higher. Though we thoroughly test, and naturally would be very thourough prior to introduction of this feature - there is a danger of problems slipping through and affecting users.
+1. Files written in newer Mantid versions wont work with v < 4 by default. 
+1. I'm not aware of anyone inspecting the `instrument_xml` directly in the output file, but this would be harder (see Nexus Constructor Tool)
+1. There would be a loss of instrument parameters as described. I'm not aware of anyone using these, but they would dissapear.
+1. Harder to compare versions between files.
+
+## 2. Explicit Replace instrument_xml in `SaveNexusProcessed`
+
+Same as above, but with the default to be writing out old-style files and opt-in for new format.
+
+**Pros**
+
+1. Potential for breaking changes low.
+1. Easy to compare files between versions of Mantid without thinking
+
+**Cons**
+
+1. Users need to be made aware of the switch
+1. Users would not necessarily take advantage of this. More of a pain at facilities like ESS where the default would be to adopt.
+1. Once parameter defaults have been set, very hard (almost impossible - see LoadInstrument rewrite spectrum map) to change them. 
+
+
+## 3. Side-by-Side Algorithms
+
+Refactor such that the common code of SaveNexusProcessed (most of it) is in a shared class/module. Keep `SaveNexusProcessed` exactly as it is and introduce a new algorithm which saves the new format.
+
+**Pros**
+
+* Users do not need to be aware of defaults at the property level for new usage
+* No breaking for old scripts
+
+**Cons**
+
+* SaveNexus is a natural name for what we want to do in both cases
+* Takes more time to refactor and create new algorithm
+
+## Notes on the Instrument Service 
+
+The instrument service allows Mantid to cache instruments and prevent reload. Both the `InstrumentDefinitionParser` and the `NexusGeometryParser` have hash mechanism to compare instruments based on content and therefore prevent uncessary reload.
```

Comment:
This is good and certainly alleviates worries about over using memory for identical copies of things.

Reply:
Well yes and this will help most of the situations, but not all.

This is because as we seperated the "base" instrument from the alterations (calibration, or movement) then the "base" instrument would be the same and then shared, even if the movements or calibrations were different.

This approach will help for non-moving and non-calibrated instruments, but not for those that move, or have different calibrations.
While most of the time runs in the same Mantid session will be from the same calibration, we will have occaisions when to runs from different cycles are compared, and this could cause significant additional load for the larger instruments.  What will confuse our users more is it will be hard to understand when mantid can and cnanot share the instrument geometry.

Reply:
@NickDraper I'm confused by this - I'm not clear on how your comments relate to the file format, you seem to be discussing in-memory "base instruments" here. I'm not clear on how the IDFs do a better/different job than what is being proposed here.